src/rp2040/armcm_canboot.c

```diff
@@ -0,0 +1,135 @@
+// CanBoot specific entry code for ARM Cortex-M vector table and bootup
+//
+// Copyright (C) 2019  Kevin O'Connor <kevin@koconnor.net>
+//
+// This file may be distributed under the terms of the GNU GPLv3 license.
+
+#include <string.h> // memcpy
+#include "generic/armcm_boot.h" // DECL_ARMCM_IRQ
+#include "autoconf.h" // CONFIG_MCU
+#include "board/internal.h" // SysTick
+#include "board/irq.h" // irq_disable
+#include "board/misc.h" // get_bootup_code
+#include "command.h" // DECL_CONSTANT_STR
+
+// Export MCU type
+DECL_CONSTANT_STR("MCU", CONFIG_MCU);
+
+// Symbols created by armcm_link.lds.S linker script
+extern uint32_t _data_start, _data_end, _data_flash;
+extern uint32_t _bss_start, _bss_end, _stack_start;
+extern uint32_t _stack_end;
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored  "-Warray-bounds"
+
+uint64_t
+get_bootup_code(void)
+{
+    uint64_t *req_code = (void*)&_stack_end;
+    return *req_code;
+}
+
+void
+set_bootup_code(uint64_t code)
+{
+    uint64_t *req_code = (void*)&_stack_end;
+    *req_code = code;
+    barrier();
+}
+
+#pragma GCC diagnostic pop
+
+// Helper function to read area of flash
+void
+application_read_flash(uint32_t address, uint32_t *dest)
+{
+    memcpy(dest, (void*)address, CONFIG_BLOCK_SIZE);
+}
+
+// Check if the application flash area looks valid
+int
+application_check_valid(void)
+{
+    uint32_t *app = (void*)CONFIG_APPLICATION_START;
+    return *app != 0 && *app != 0xffffffff;
+}
+
+// Jump to the main application (exiting the bootloader)
+void
+application_jump(void)
+{
+    irq_disable();
+    set_bootup_code(REQUEST_START_APP);
+    NVIC_SystemReset();
+}
+
+static void
+start_application(void)
+{
+    set_bootup_code(0);
+    uint32_t *vtor = (void*)CONFIG_APPLICATION_START;
+    SCB->VTOR = (uint32_t)vtor;
+    asm volatile("MSR msp, %0\n    bx %1" : : "r"(vtor[0]), "r"(vtor[1]));
+}
+
+void __noreturn __visible
+__attribute__((used, section(".reset_handler_flash.reset_handler_stage_two")))
+reset_handler_stage_two(void)
+{
+    // Copy global variables from flash to ram
+    uint32_t count = (&_data_end - &_data_start);
+    for(int i = 0; i < count; i++) {
+        (&_data_start)[i] = (&_data_flash)[i];
+        barrier();
```

Comment:
At this point it is not safe to call anything, cause we did not copied code yet. So barrier() here to prevent compiler from use memcpy and memset.